Remove dead code throughout code base

alexdewar opened the pull request:
# Description

Here's another sledgehammer :wink:.

Just a note on the diff: it looks like I've removed a lot more code than I have based on number of lines changed but that's because I removed the rather large `.graphml` file.

I'm opening this as a draft PR because I've done this *very* aggressively and it is possible I've taken out something you actually want. Initially I noticed some unused functions in `Autocorpus.py` and thought about just removing those, but realised there were other unused functions dotted around so figured it was better to try to clear all the baggage at once. I've tested it manually by running `auto-corpus` from the command line using the sample data files we have and the tests pass. Hopefully clearing this out should make future development a bit easier.

I've assumed -- perhaps wrongly -- that users will mostly be interacting with AC through the command-line interface or by invoking the same functionality in a Python script (i.e. just using the Python package to run processing over a file or files rather than because they want to access the various helper functions in the package). Let me know if I've got this wrong!

There was a big chunk of code in the `utils.py` file related to a DAG model (whatever that is), which I removed because it doesn't seem that that function is actually called from anywhere else in the codebase. Removing it also meant we could drop the dependency on the `networkx` package and the `.graphml` file. It does seem like an obvious case of functionality that isn't replicated elsewhere though, so let me know if you want to keep it. If we *do* want it, I'd suggest we find a way to let users access this functionality via the command-line interface (as with the other bits I've stripped out).

I used the [`vulture`](https://pypi.org/project/vulture/) package to find all the unused code/variables.

Closes #150.

## Type of change

- [ ] Documentation (non-breaking change that adds or improves the documentation)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (non-breaking, back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (whatever its nature)

## Key checklist

- [x] All tests pass (eg. `pytest`)
- [ ] The documentation builds and looks OK (eg. `mkdocs`)
- [x] Pre-commit hooks run successfully (eg. `pre-commit run --all-files`)

## Further checks

- [ ] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added or an issue has been opened to tackle that in the future. (Indicate issue here: # (issue))

Thomas-Rowlands commented:
LGTM, test passes locally

AdrianDAlessandro reviewed (commented):
I'm very happy to remove all of this... However we should get input from the others before cutting these out. Particularly the unused functions and methods.

alexdewar commented on autocorpus/Autocorpus.py:
```diff
@@ -52,58 +37,6 @@ def __soupify_infile(self, fpath):
         except Exception as e:
             print(e)
 
-    def __clean_text(self, result: dict) -> dict:
```
Yeah... who knows why it isn't :shrug:

AdrianDAlessandro commented:
@jmp111 @Antoinelfr We should get your input on this before we remove some of these functions. Hard to tell if they're not used because they're not useful, or if it's because there was an error made and they _should_ be used.

AdrianDAlessandro commented on autocorpus/__main__.py:
```diff
@@ -38,14 +35,9 @@
-group.add_argument(
+parser.add_argument(
     "-c", "--config", type=str, help="filepath for configuration JSON file"
 )
```
We should keep this mutually exclusive group because it's being used in #149

alexdewar replied:
Oops! I'll put it back

AdrianDAlessandro reviewed (approved):
Poetry lockfile has a conflict. Sort that and then we can merge.

Also, following discussion this morning. Can you open an issue about the DAG being used later?

alexdewar commented:
I've opened an issue about the DAG model: #169 

I've just noticed that the `--trained_data` option and associated variables are also currently unused. Is there a plan to fix this? I'm a bit confused by this. It seems like at least some of the code in `tableimage.py` is used, but I thought the point of that code was to use the OCR training data, so I'm not sure what it's actually doing. If no one is using this, we could give it the same treatment as the DAG model, i.e. remove it and add an issue for reintroducing the functionality if people want it (that would also let us remove the rather large OpenCV dependency). If we do want to keep the code, I think we should try to add this functionality to the command-line program soon though, otherwise it will just bitrot and be an unnecessary maintenance burden.

Thoughts?

AdrianDAlessandro commented:
I say remove it and open an issue like we've done for the DAG

alexdewar commented:
It seems likely that no one has used this functionality for years as there's no way to do so through the command line interface. I'll rip it out. I've opened an issue for reintroducing it: #173

alexdewar commented:
Done. I was able to remove the `pytesseract`, OpenCV and `filetype` dependencies after doing that.

Unrelated change: I noticed that `python-levenshtein` was listed as a dependency in `pyproject.toml`, but it's actually only needed as an optional extra for `fuzzywuzzy`. I've updated `pyproject.toml` to reflect this, which should help avoid version conflicts.

We previously had an indirect dependency on `numpy` via OpenCV, but now that I've removed it, we were getting an error because @Antoinelfr's `parse_xml.py` script uses `numpy`. I considered re-adding `numpy` as a dependency, but there's literally only one line of code that uses it, so I just replaced it with a non-`numpy` alternative. Hope that's ok.

This is the final tally:

```
 autocorpus/Autocorpus.py     |   72 --
 autocorpus/DAG_model.graphml | 2328 ----------------------------------------
 autocorpus/__main__.py       |   51 +-
 autocorpus/abbreviation.py   |    2 +-
 autocorpus/parse_xml.py      |    5 +-
 autocorpus/table.py          |   94 --
 autocorpus/tableimage.py     |  397 -------
 autocorpus/utils.py          |  126 ---
 poetry.lock                  | 1118 ++++++++-----------
 pyproject.toml               |    7 +-
 trainedData/all.traineddata  |  Bin 684297 -> 0 bytes
 trainedData/eng.traineddata  |  Bin 15400601 -> 0 bytes
 trainedData/num.traineddata  |  Bin 169030 -> 0 bytes
 trainedData/sym.traineddata  |  Bin 564178 -> 0 bytes
 14 files changed, 476 insertions(+), 3724 deletions(-)
```

Aside from that, we've also been able to drop four direct dependencies. Ahhhhh. I love the smell of deleted code in the morning.

AdrianDAlessandro commented on autocorpus/parse_xml.py:
```diff
@@ -648,7 +647,7 @@ def find_parent_titles(title_element: Tag | NavigableString) -> list[str]:
 
                 # Remove duplicate closing tags and create a list of unique matches
-                good_matches = list(np.unique(matches))
+                good_matches = list(dict.fromkeys(matches))
```
Would a `set` have made more sense here? I'm a bit thrown by making a dictionary and immediately casting it to a list

alexdewar replied:
That was my first thought, but the problem is that `set`s are unordered so your matches will end up jumbled up. Might not be a problem here, but I didn't want to change the behaviour.

Credit/blame for the `dict` hack goes to SO: https://stackoverflow.com/questions/12897374/get-unique-values-from-a-list-in-python
